Limit fullscreen to window only

quxs opened the pull request:
#3213 Adds `content.desktop_fullscreen` config option to enable/disable fullscreen covering the entire desktop.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3285)
<!-- Reviewable:end -->

jgkamat commented:
I'm not sure if this is intended behavior, but I wanted to point it out.

1. set `content.desktop_fullscreen` to false
2. html fullscreen a window
3. Have your WM fullscreen qb
4. Un-fullscreen the html video
5. qb will stop being fullscreen. (which is a bit unexpected for me)

I think that manually making qb fullscreen will be a common gesture will be common (if people actually want to have a full-screen video), so it's important to get this 'right'. I don't mind either way though.

quxs commented:
Good catch! I added a fix that only changes the window to it's previous state if `desktop_fullscreen` is `true`.

 I also added a check when leaving fullscreen to see if `desktop_fullscreen` has changed from `true` to `false` while in html fullscreen. This handles the following issue:
1. Start with window not set to fullscreen by WM.
2. Set `desktop_fullscreen = true`
2. html fullscreen
3. Set `desktop_fullscreen = false`
4. Leaving html fullscreen leaves the qutebrowser window in fullscreen

It's a very rare case and I'm not sure if it's worth handling live config changes. I can remove it if needed.

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -259,6 +259,12 @@ content.default_encoding:
     The encoding must be a string describing an encoding such as _utf-8_,
     _iso-8859-1_, etc.
 
+content.desktop_fullscreen:
```
This setting name seems kind of confusing to me, as a fullscreen being full-screen (duh 😆) is kind of the normal thing I think people would expect. What about naming it `windowed_fullscreen` instead and inverting the meaning?

The-Compiler commented on qutebrowser/mainwindow/mainwindow.py:
```diff
@@ -492,10 +492,13 @@ def _connect_signals(self):
         if on:
-            self.state_before_fullscreen = self.windowState()
-            self.showFullScreen()
+            self.window_state_before_fullscreen = self.windowState()
```
If you really want to rename this, you'll also need to adjust the name in `__init__` and in other places it's used (`browser/commands.py`).

The-Compiler commented on qutebrowser/mainwindow/mainwindow.py:
```diff
@@ -492,10 +492,13 @@ def _connect_signals(self):
-            self.state_before_fullscreen = self.windowState()
-            self.showFullScreen()
+            self.window_state_before_fullscreen = self.windowState()
+            self.config_state_before_fullscreen = config.val.content.desktop_fullscreen
```
You'll also need to initialize this in `__init__` correctly. I'm also not sure if it makes sense to name those so similarly, as they are two completely different things.

quxs replied:
I've gone ahead and removed that second variable and check because I'm now thinking it's not really unexpected behaviour.

The-Compiler commented:
Hmm, travis doesn't like you 😢

quxs commented:
@The-Compiler Do you have any guidance on fixing the CI issues? I'm not sure where to look.

jgkamat commented:
@quxs, sorry, those are actually errors from my branch merged yesterday, I suspect that your code is fine, and that if travis rebuilds it will give you a :+1:

The-Compiler commented:
Thanks!
